### DISCUSSION TITLE
avanti, indietro & attraversamento

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#7] [Esercizi A-L]
Dato un grafo **diretto** $G = (V, E)$ scrivere lo pseudocodice di un algoritmo che trova gli archi **in avanti**, all'**indietro** e di **attraversamento** in $O(|V|+|E|)$

**Comment by Frisbon** [comment]
> Si suppone che venga passato un grafo **fortemente connesso** (e quindi posso partire da un nodo qualsiasi) oppure un grafo **ordinato** insieme alla sua **radice** perché se (vedi esempio) dovessi partire da C, non potrei mai raggiungere gli altri nodi.

![WhatsApp Image 2024-03-13 at 12 46 30_579f9fc2](https://github.com/sapienzastudentsnetwork/progalg2324/assets/84731341/2264380e-a992-4a2c-8015-0483c0e08a40)
### Uso due funzioni separate (opzionale):

- **Scan_times:** DFS che si salva gli archi e i tempi di chiusura/apertura.
    
![image](https://github.com/sapienzastudentsnetwork/progalg2324/assets/84731341/a826b797-d534-4f08-b5bf-c0ed4d14f7c9)
    
- **Classifica_archi:** utilizza 3 insiemi *Back, Forward* e *Cross* che saranno il mio output.
    
![image](https://github.com/sapienzastudentsnetwork/progalg2324/assets/84731341/76c1dcab-c645-43aa-af2c-b3b22a5e63d1)

**Reply by CasuFrost** [reply]
Non ho visto la soluzione, ma solamente la premessa, se dovessi iniziare da $c$ non ci sarebbero problemi, semplicemente non ci sarebbero archi in avanti, all'indietro e di attraversamento. Questi ultimi sono proprietà dell'esecuzione di una DFS, non di un grafo

**Comment by daromacarlo** [comment]
```c
def DFS_CON_CONTATORE_rec(G, CC, counter, x, vis, albero){
	//CC = lista di liste da due elementi dove il primo elemento indica l'istante di apertura del nodo mentre il secondo elemento indica l'istante di chiusura del
	
	vis[x] = 1
	x.CC[t] = counter
	for w in x.adiacenti(){
		if vis[w] == 0{
			DFS_CON_CONTATORE(G, CC, counter++, w, vis)
			albero += [(x, w)]
		}
	}
	x.CC[T] = counter
	}
}
return CC, albero

//O(n+m)

def Identifica_archi(G)
	albero = []
	CC, albero = DFS_CON_CONTATORE_rec(G, CC = [[t1, T1],...,[tn, Tn]], counter = 0, x = random(V(G)), vis = [0,...,0], albero = [])
	annanze = [] 
	arrete = []
	attraversamento = []
	for (x,y) in E(G) and not in albero{											//per ogni arco in E(G)
		if (x.CC[t] < y.CC[t]) && (x.CC[T] >= y.CC[T]){								//se apro e chiudo y tra l'apertura e la chiusura di x
			annaze += [(x,y)]
		}
		if (y.CC[t] < x.CC[t]) && (y.CC[T] >= x.CC[T]){								//se apro e chiudo x tra l'apertura e la chiusura di y
			arrete += [(x,y)]
		else{
			attraversamento += [(x,y)]												//altrimenti
		}
	}
	return annanze, arrete, attraversamento

```

**Comment by daromacarlo** [comment]
dfs con contatore (versione non ricorsiva)
```c
def DFS_CON_CONTATORE(G, CC, counter, x, vis, albero)
	
	stack = []
	stack.push(x)
	while stack is not empty{
		v = stack.top()
		v.CC[t] = counter
		counter ++
		if v.adiacenti[0] == 0{
			w = v.adiacenti[0]
			stack.push(w)
		}
		else{
			v.adiacenti.remove[0]
		}
		v.CC[T] = contatore
		stack.pop()
	}
	return CC

//O(N+M)
```
